lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/HeartBeatManager.cs

```diff
@@ -166,17 +169,31 @@ public void Send(EvaluatorHeartbeatProto evaluatorHeartbeatProto)
                     _queuedHeartbeats.Enqueue(evaluatorHeartbeatProto);
                     LOGGER.Log(Level.Error, string.Format(CultureInfo.InvariantCulture, "Sending heartbeat to driver experienced #{0} failure. Hearbeat queued as: [{1}]. ", _heartbeatFailures, evaluatorHeartbeatProto), e);
 
-                    if (_heartbeatFailures >= _maxHeartbeatRetries)
+                    if (_driverConnection.Get() is MissingDriverConnection)
                     {
-                        LOGGER.Log(Level.Warning, "Heartbeat communications to driver reached max of {0} failures. Driver is considered dead/unreachable", _heartbeatFailures);
-                        LOGGER.Log(Level.Info, "=========== Entering RECOVERY mode. ===========");
-                        ContextManager.HandleDriverConnectionMessage(new DriverConnectionMessageImpl(DriverConnectionState.Disconnected));
-
-                        LOGGER.Log(Level.Info, "instantiate driver reconnect implementation: " + _driverConnection);
-                        _evaluatorSettings.OperationState = EvaluatorOperationState.RECOVERY;
-
-                        // clean heartbeat failure
-                        _heartbeatFailures = 0;
+                        if (_heartbeatFailures >= _maxHeartbeatRetriesForNonRecoveryMode)
+                        {
+                            LOGGER.Log(Level.Error, "Have encountered {0} heartbeat failures. There is no IDriverConnection implemented for HA.", _heartbeatFailures);
+                            throw;
```

Comment:
If we `throw;` here, we'll be throwing a random network exception which happened to be the last one in the series of failures. I'd create a dedicated exception with error message which says something along the lines of "Limit of heartbeat sending failures exceeded, driver reconnect logic not implemented, failing evaluator"